Complete migration from Easy_logging to Logs

mjambon opened the pull request:
This supports tags for filtering (see `Logs_.mli`). Tags are logged and can be filtered using the environment variable `LOG_TAGS` (or with `grep` for messages that don't contain newlines). Here's what the output looks like for

```
semgrep -e hello -l python <(echo hello) --debug
```

![image](https://github.com/semgrep/semgrep/assets/343265/70ddaf72-7dd4-4ca6-b0e2-b4a6b948f8ca)

With filtering:
```
$ LOG_TAGS=Analyze_rule,Core_scan,asdf semgrep -e hello -l python <(echo hello) --debug
```
![image](https://github.com/semgrep/semgrep/assets/343265/10f3a769-3f18-4163-8741-d1e9a343a2b8)

Semgrep-pro PR: https://github.com/semgrep/semgrep-proprietary/pull/1282

Commit:
Enable logging with Logs in semgrep-core + tag support.
This adds support for tagging and filtering log messages with the Logs
library.

Commit:
Add Logs_.sdebug etc.

Commit:
Autoreplace calls to the old logger with calls to the new one
+ minor manual adjustments
Command: semgrep -c logger.yaml -a
Contents of logger.yaml:
rules:
- id: debug
  pattern: logger#debug $...ARGS
  fix: Logs.debug (fun m -> m ~tags $...ARGS)
  message: logger call
  languages: [ocaml]
  severity: ERROR
- id: trace
  pattern: logger#trace $...ARGS
  fix: Logs.debug (fun m -> m ~tags $...ARGS)
  message: logger call
  languages: [ocaml]
  severity: ERROR
- id: info
  pattern: logger#info $...ARGS
  fix: Logs.info (fun m -> m ~tags $...ARGS)
  message: logger call
  languages: [ocaml]
  severity: ERROR
- id: warning
  pattern: logger#warning $...ARGS
  fix: Logs.warn (fun m -> m ~tags $...ARGS)
  message: logger call
  languages: [ocaml]
  severity: ERROR
- id: error
  pattern: logger#error $...ARGS
  fix: Logs.err (fun m -> m ~tags $...ARGS)
  message: logger call
  languages: [ocaml]
  severity: ERROR
- id: flash
  pattern: logger#flash $...ARGS
  fix: Logs.err (fun m -> m ~tags $...ARGS)
  message: logger call
  languages: [ocaml]
  severity: ERROR

- id: sdebug
  pattern: logger#sdebug
  fix: Logs_.sdebug ~tags
  message: logger call
  languages: [ocaml]
  severity: ERROR
- id: sinfo
  pattern: logger#sinfo
  fix: Logs_.sinfo ~tags
  message: logger call
  languages: [ocaml]
  severity: ERROR

- id: ldebug
  pattern: |
    logger#ldebug (lazy $STR)
  fix: |
    Logs.debug (fun m -> m ~tags "%s" $STR)
  message: logger call
  languages: [ocaml]
  severity: ERROR
- id: linfo
  pattern: |
    logger#linfo (lazy $STR)
  fix: |
    Logs.info (fun m -> m ~tags "%s" $STR)
  message: logger call
  languages: [ocaml]
  severity: ERROR

- id: init
  pattern: |
    let logger = Logging.get_logger [ __MODULE__ ]
  fix: |
    tags = Logs_.create_tags [ __MODULE__ ]
  message: logger call
  languages: [ocaml]
  severity: ERROR

Commit:
Add support for filtering tags using an environment variable

Commit:
Remove dependency on Easy_logging and port -log_to_file to Logs

Commit:
Merge remote-tracking branch 'origin/develop' into martin/logs

aryx commented on languages/c/ast/ast_c_build.ml:
```diff
@@ -17,7 +17,7 @@ open Either_
 module A = Ast_c
 
-let logger = Logging.get_logger [ __MODULE__ ]
+let tags = Logs_.create_tags [ __MODULE__ ]
```
What is the difference between Logs tags and Logs src?
I've seen logs calls "tagged" with different src.

mjambon replied:
I didn't know about sources or what to do with them. The [documentation](https://erratique.ch/software/logs/doc/Logs/Src/index.html) says it's something like the current library or module but I don't see the point in making it different from ordinary tags. The point of tags, for me, is to be able to filter the logs and show what's relevant to the issue I'm troubleshooting. Having the name of the module as a tag is a good starting point but it's not necessary. We should feel free to add tags that describe a particular flow affecting various modules.

aryx commented on languages/c/ast/ast_c_build.ml:
```diff
@@ -77,7 +77,8 @@ let rec ifdef_skipper xs f =
           match ifdef with
           | Ifdef tok -> (
-              logger#info "skipping: %s" (Tok.content_of_tok tok);
+              Logs.info (fun m ->
```
Like I said in Slack, most of our logger#info should be converted in Logs.debug,
because Logs.info can be seen by regular users when using --verbose.

mjambon replied:
I'll refine the recommendations in `Logs_.mli` as they become clearer to us.

aryx commented on languages/c/ast/ast_c_build.ml:
```diff
@@ -627,7 +628,7 @@ and expr env e =
   | IdSpecial (SizeOf, tk) -> A.IdSpecial (SizeOf, tk)
   | ConstructedObject (_, _) ->
-      logger#error "BUG PARSING LOCAL DECL PROBABLY";
+      Logs.err (fun m -> m ~tags "BUG PARSING LOCAL DECL PROBABLY");
```
This should be Logs.err, like you did, but then that means every semgrep users will now see those errors ...

I think that we should configure logging so that any Logs.err or Logs.info with a ~tags should be masked and shown only when you use --debug really.

mjambon replied:
My conclusion, as mentioned on Slack, is that no log message should be seen on stdout or stderr by the command-line user unless they turn it on (or run a server loop). In the end, these error-level messages are debugging helps just like all the other levels.

aryx replied:
yes but then you should change this one to Logs.debug, because Logs.err will get displayed to the user I think.

aryx commented on libs/commons/Logs_.ml:
```diff
@@ -50,6 +50,63 @@ let default_skip_libs =
+(* String tags *)
+(*****************************************************************************)
+(*
+   The interface of the Logs.Tag module is complicated. Here we assume
```
This daniel buenzli is making good libraries (Logs, Cmdliner, Fpath), but they are not super easy to use at first ... the learning curve is deep.

aryx commented on libs/commons/Logs_.ml:
```diff
@@ -71,10 +128,41 @@ let pp_sgr ppf style =
+
+(* log reporter
+
+   This code was copy-pasted and derived from the example in the Logs library.
```
I agree

aryx commented on libs/commons/Logs_.mli:
```diff
@@ -1,24 +1,30 @@
 
       - App, this level can be used for the standard output or console
         of an application. It should never be used by libraries.
+        Semgrep: We don't use it.
```
we do actually; git grep Logs.app return many matches

mjambon replied:
got it.

aryx commented on libs/commons/Logs_.mli:
```diff
@@ -32,9 +38,40 @@ val default_skip_libs : string list
 (* Setup the Logs library. This call is necessary before any logging
    calls, otherwise your log will not go anywhere (not even on stderr).
+
+   'require_one_of_these_tags': if a list of tags is provided, at least one
```
missing 'of'

aryx commented on libs/commons/Logs_.mli:
```diff
@@ -32,9 +38,40 @@ val default_skip_libs : string list
+   to make semgrep run at the debug level but only show lines tagged with
+   'Match_rules' or 'Core_scan'.
+
+     $ LOG_TAGS=Match_rules,Core_scan semgrep -e 'Obj.magic' -l ocaml --debug
```
nice

aryx reviewed (approved):
We need to make sure that by default, and also when using --verbose, all those new Logs.info Logs.err Logs.warn are actually not displayed; they should be displayed only when we use --debug, otherwise suddently users will get far more messages.

Commit:
Restore the behavior of Logs.app: print a line on stderr with no decorations.

Commit:
semgrep-core revert: log nothing by default

Commit:
Clarify Logs_ doc

Commit:
Update comment

Commit:
Merge remote-tracking branch 'origin/develop' into martin/logs
Oops. This also includes changing the log level for many of
freshly-translated calls to Logs.info so as to reduce the noise of the
--verbose output.

Commit:
Remove retired code

Commit:
Move cryptic warnings to the debug level

Commit:
Attention: Any log message that can't be understood without context
will be moved to the Debug level!

Commit:
Don't print a spurious trailing space
(causes problems in the osemgrep implementation and has no reason to be
anyway)

Commit:
Update snapshots (removing trailing space)

Commit:
Add a module Std_msg to hold the global text highlighting/color options.
Logging uses this setting by default but it can be overridden.
Osemgrep doesn't use Std_msg yet and maybe shouldn't until we're
fully migrated away from pysemgrep because we have to reproduce its
output exactly.

Commit:
Fix some snapshots

Commit:
Update snapshots getting rid of the trailing space after file name

Commit:
Mask a bunch of material in --debug output

Commit:
More masking

Commit:
Yet another snapshot mask

Commit:
Declare calendar (CalendarLib module) as an opam dependency
since it's no longer indirectly installed by the logging library but
still used directly in semgrep.

Commit:
Add another opam dependency that was missing

Commit:
Ugh. Regenerate opam file.

Commit:
Merge remote-tracking branch 'origin/develop' into martin/logs

Commit:
Merge remote-tracking branch 'origin/develop' into martin/logs

Commit:
Merge remote-tracking branch 'origin/develop' into martin/logs

Commit:
Clarify confusing test output

Commit:
Fix bug in Std_msg (highlight_setting was not set properly)

Commit:
Update test snapshots

Commit:
Eliminate custom solution to redirect and check logs
One test with a mystery 'exit 2' exception was marked as XFAIL
(test_publish)

Commit:
Fix annoying differences in escape sequences between pysemgrep and osemgrep

Commit:
Merge remote-tracking branch 'origin/develop' into martin/logs

aryx commented on cli/tests/conftest.py:
```diff
@@ -331,8 +331,8 @@ def as_snapshot(self, mask: Optional[Maskers] = None):
         ):
-            del sections["stdout - color"]
-            del sections["stderr - color"]
+            sections["stdout - color"] = "<same as above: stdout - plain>"
```
a comment about those sections would be useful.

aryx replied:
Would have been nice to also put that in a separate PR because this PR is already big.

mjambon replied:
> Would have been nice to also put that in a separate PR because this PR is already big.

PRs are too heavy. If CI checks took less than 5 min, then yes, I'd make another PR. Right now, it's this or doing nothing at all, with people generally erring toward the latter.

aryx commented on cli/tests/e2e/snapshots/test_metrics/test_metrics_payload/pro_flag0/metrics-payload.json:
```diff
@@ -56,7 +56,7 @@
       }
     ],
-    "maxMemoryBytes": 48693248,
+    "maxMemoryBytes": 42336256,
```
This one is annoying. I've seen it changed a few times; can't we just mask it?

mjambon replied:
https://github.com/semgrep/semgrep/pull/9759

aryx commented on cli/tests/e2e/snapshots/test_output/test_debug_experimental_rule/results.txt:
```diff
@@ -1,5 +1,5 @@
 === command
-SEMGREP_USER_AGENT_APPEND="pytest" SEMGREP_SETTINGS_FILE="<MASKED>" SEMGREP_VERSION_CACHE_PATH="<MASKED>" SEMGREP_ENABLE_VERSION_CHECK="0" SEMGREP_SEND_METRICS="off" semgrep --debug --strict --config rules/experiment/experiment.yaml targets/experiment/experiment.py
+SEMGREP_USER_AGENT_APPEND="pytest" SEMGREP_SETTINGS_FILE="<MASKED>" SEMGREP_VERSION_CACHE_PATH="<MASKED>" SEMGREP_ENABLE_VERSION_CHECK="0" SEMGREP_SEND_METRICS="off" semgrep --debug --strict --config <MASKED> <MASKED>
```
why the mask here?

mjambon replied:
Does it hurt? There were a bunch of very specific masking rules. I replaced them by a simpler one that masks all paths because it's easier to maintain.

aryx commented on cli/tests/e2e/snapshots/test_output/test_debug_experimental_rule/results.txt:
```diff
@@ -64,7 +140,7 @@ Files skipped:
 
   Skipped by .semgrepignore:
-  - https://semgrep.dev<MASKED>
+  - https:/<MASKED>
```
again the MASK seems wrong.

mjambon replied:
same remark as above.

aryx commented on libs/commons/Std_msg.mli:
```diff
@@ -0,0 +1,73 @@
+   The 'Logs_' module which depends on this one because logging
+   is done on stderr by default.
+
+   TODO: use this module for printing all user-facing messages.
```
That was the goal of the CapConsole.ml module.

mjambon replied:
then we can call it `Console` or `Console_` since it doesn't do anything special regarding capabilities (if we're going to use them everywhere, we're not going to prefix every module with `Cap`).

aryx commented on src/core_cli/Core_CLI.ml:
```diff
@@ -680,9 +680,13 @@ let main_no_exn_handler (caps : Cap.all_caps) (sys_argv : string array) : unit =
+    ~level:
+      (if config.debug then Some Debug
+         (* else if config.verbose then Some App *)
+       else None)
```
why we changed from Some Info to None now?

mjambon replied:
Probably because semgrep-core suddenly started printing a lot of messages and the simplest way for me to deal with it was to suppress them. A sane thing to do would be to have a command-line option (and an environment variable) for choosing the log level. Right now, it's all or nothing. My comment `(* else if config.verbose then Some App *)` is a suggestion that we support `-verbose` like `semgrep` supports `--verbose`.

But this is semgrep-core... I don't know and don't care much about log levels. I'm fine with `-debug` and `LOG_TAGS=...` if necessary.

I'll try to see what happens if we default to Info in a separate PR.

mjambon replied:
Yes, with Info, we're getting a bunch of log messages at all levels, including ERROR. See a sample semgrep-core run below where the log level is Info:

```
~/semgrep $ semgrep-core -l py -e x cli/tests/ 
[00.00][INFO](cli, Core_CLI): Executed as: semgrep-core -l py -e x cli/tests/
[00.00][INFO](cli, Core_CLI): Version: semgrep-core version: 1.60.1
[00.05][WARNING](Core_command): -fast does not work with -f/-e, or you need also -json
cli/tests/e2e-pro/targets/autofix/autofix.py:5
   inputs[x] = 1
cli/tests/e2e-pro/targets/autofix/autofix.py:6
   if inputs[x + 1] == True:
cli/tests/e2e-pro/targets/autofix/python-assert-statement.py:7
 x = "abc"
cli/tests/e2e-pro/targets/autofix/python-ranges.py:8
 foo(x[1])
cli/tests/e2e-pro/targets/autofix/python-ranges.py:9
 foo(x[1, 2])
[00.10][ERROR](Pfff_or_tree_sitter): exn (Parsing_error.Syntax_error (cli/tests/e2e-pro/targets/bad/invalid_python.py:1:10 "\n    ")) with Pfff parser
[00.10][ERROR](Parse_tree_sitter_helpers): Partial errors returned by Tree-sitter parser
node type: ERROR
children: [
  "def"
  identifier
  parameters
  identifier
  "="
  identifier
  "."
  identifier
  "("
  string
  ")"
]
Unrecognized construct
[00.10][ERROR](Pfff_or_tree_sitter): partial errors (1) with TreeSitter parser
WARNING: fail to fully parse cli/tests/e2e-pro/targets/bad/invalid_python.py
cli/tests/e2e-pro/targets/basic/stupid.py:8
     x == x
cli/tests/e2e-pro/targets/basic/stupid.py:8
     x == x
cli/tests/e2e-pro/targets/basic/stupid.py:12
 assertTrue(x == x)
cli/tests/e2e-pro/targets/basic/stupid.py:12
 assertTrue(x == x)
cli/tests/e2e-pro/targets/ci/foo.py:9
     x == x  # nosemgrep
cli/tests/e2e-pro/targets/ci/foo.py:9
     x == x  # nosemgrep
cli/tests/e2e-pro/targets/ci/foo.py:15
     x == 5
cli/tests/e2e-pro/targets/match_based_id/before/join.py:2
 x = z+z
cli/tests/e2e-pro/targets/match_based_id/before/join.py:3
 if z == x:
cli/tests/e2e-pro/targets/message_interpolation/pattern_either_basic.py:2
     print(x)
cli/tests/e2e-pro/targets/message_interpolation/propagated_constant.py:3
        x = 0
cli/tests/e2e-pro/targets/message_interpolation/propagated_constant.py:4
        return x == 0
cli/tests/e2e-pro/targets/multiline/stupid.py:11
 x = """
cli/tests/e2e-pro/targets/taint/taint.py:2
     x = 'test string format {}'.format('foo')
cli/tests/e2e-pro/targets/version-constraints/x.py:1
 x = "hello"
cli/tests/e2e-pro/test_api.py:48
     x = subprocess.run(
cli/tests/e2e-pro/test_api.py:58
     assert x.stdout == ""
cli/tests/e2e-pro/test_api.py:59
     assert x.stderr == ""
cli/tests/e2e/targets/autofix/autofix.py:5
   inputs[x] = 1
cli/tests/e2e/targets/autofix/autofix.py:6
   if inputs[x + 1] == True:
cli/tests/e2e/targets/autofix/python-assert-statement.py:7
 x = "abc"
cli/tests/e2e/targets/autofix/python-ranges.py:8
 foo(x[1])
cli/tests/e2e/targets/autofix/python-ranges.py:9
 foo(x[1, 2])
[00.21][ERROR](Pfff_or_tree_sitter): exn (Parsing_error.Syntax_error (cli/tests/e2e/targets/bad/invalid_python.py:1:10 "\n    ")) with Pfff parser
[00.21][ERROR](Parse_tree_sitter_helpers): Partial errors returned by Tree-sitter parser
node type: ERROR
children: [
  "def"
  identifier
  parameters
  identifier
  "="
  identifier
  "."
  identifier
  "("
  string
  ")"
]
Unrecognized construct
[00.21][ERROR](Pfff_or_tree_sitter): partial errors (1) with TreeSitter parser
WARNING: fail to fully parse cli/tests/e2e/targets/bad/invalid_python.py
cli/tests/e2e/targets/basic/stupid.py:8
     x == x
cli/tests/e2e/targets/basic/stupid.py:8
     x == x
cli/tests/e2e/targets/basic/stupid.py:12
 assertTrue(x == x)
cli/tests/e2e/targets/basic/stupid.py:12
 assertTrue(x == x)
cli/tests/e2e/targets/ci/foo.py:9
     x == x  # nosemgrep
cli/tests/e2e/targets/ci/foo.py:9
     x == x  # nosemgrep
cli/tests/e2e/targets/ci/foo.py:15
     x == 5
cli/tests/e2e/targets/match_based_id/before/join.py:2
 x = z+z
cli/tests/e2e/targets/match_based_id/before/join.py:3
 if z == x:
cli/tests/e2e/targets/message_interpolation/pattern_either_basic.py:2
     print(x)
cli/tests/e2e/targets/message_interpolation/propagated_constant.py:3
        x = 0
cli/tests/e2e/targets/message_interpolation/propagated_constant.py:4
        return x == 0
cli/tests/e2e/targets/multiline/stupid.py:11
 x = """
cli/tests/e2e/targets/taint/taint.py:2
     x = 'test string format {}'.format('foo')
cli/tests/e2e/targets/version-constraints/x.py:1
 x = "hello"
[00.32][ERROR](AST_to_IL): cli/tests/fixtures.py:35:8: Cannot translate Semgrep construct(s) into IL
```

mjambon replied:
done: https://github.com/semgrep/semgrep/pull/9757

aryx commented on src/osemgrep/cli_login/Test_login_subcommand.ml:
```diff
@@ -74,31 +58,28 @@ let with_fake_deployment_response return_value f =
-         with_logs
-           ~f:(fun () -> Logout_subcommand.main caps [| "semgrep-logout" |])
-           ~final:(fun res ->
-             assert (res.logs =~ ".*You are not logged in");
```
where do we check that the stderr contains "You are not logged in" ?
I see a ~normalize: [Test.mak_not_substring, 
but it's not clear why those 2 functions used together assert that the string is in the stderr.

mjambon replied:
```ocaml
let test_logout_not_logged_in caps : Testo.test =
  t ~checked_output:Stderr
    ~normalize:[ Testo.mask_not_substring "You are not logged in" ]
    __FUNCTION__
    (with_login_test_env (fun () ->
         let exit_code = Logout_subcommand.main caps [| "semgrep-logout" |] in
         assert (exit_code =*= Exit_code.ok)))
```
`~normalize:[ Testo.mask_not_substring "You are not logged in" ]` means the whole output is masked except for `You are not logged in`.

It creates a test snapshot that's `<MASKED>You are not logged in<MASKED>`. If the message disappears or becomes `You are logged out`, the test will fail and show a diff.

aryx replied:
I find the testing code a bit more difficult to read than the old code with the simple assert string is in the output.

aryx commented on src/osemgrep/cli_login/Test_login_subcommand.ml:
```diff
@@ -119,45 +100,49 @@ let fake_deployment =
+             (caps :> < Cap.stdout >)
+             [| "semgrep-logout" |]
+         in
+         (* assert (res.logs =~ ".*You are not logged in"); *)
```
what about to those assert?

mjambon replied:
This is now handled by capturing the output and comparing against the reference snapshot (in which everything is masked except for these strings we want to check).

I'll remove the asserts.

mjambon replied:
https://github.com/semgrep/semgrep/pull/9762

aryx reviewed (approved):
It's a big PR ... I have a few request for changes, but let's merge this and address them in a separate PR because I'm scared of you getting conflicts.
